### PR TITLE
Addresses #2610, insertComponent can create duplicate unique model objects

### LINKED ISSUE #2610: insertComponent can create duplicate unique model objects
m = OpenStudio::Model::Model.new
m.getSite
m.getSiteWaterMainsTemperature
m.getWeatherFile

remote = OpenStudio::RemoteBCL.new
remote.downloadComponent('f38beff0-edc6-0131-6605-48e0eb16a403')
component = remote.waitForComponentDownload()
component = component.get
osc_path = component.files("osc")[0]
vt = OpenStudio::OSVersion::VersionTranslator.new
component_object = vt.loadComponent(OpenStudio::Path.new(osc_path))
componentData = m.insertComponent(component_object.get)
puts m

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #2610
 - Also adds a bunch of convenience methods on `model` for checking for unique model object existence (without creating them).

Please read [OpenStudio Pull Requests](https://github.com/NREL/OpenStudio/wiki/OpenStudio-Pull-Requests) to better understand the OpenStudio Pull Request protocol.

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: test.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# remote.downloadComponent('d3ac66a0-acec-012f-0c6a-00ff10b04504')
+# component = remote.waitForComponentDownload()
+# component = component.get
+# osc_path = component.files("osc")[0]
+osc_path = 'C:\Users\jrobert1\BCL\d3ac66a0-acec-012f-0c6a-00ff10b04504\328940ab-bb98-4fab-a28a-c39284962d3b\files\watermains.osc'
+vt = OpenStudio::OSVersion::VersionTranslator.new
+component_object = vt.loadComponent(OpenStudio::Path.new(osc_path))
+componentData = m.insertComponent(component_object.get)
+
+puts m
```
Comment (jmarrec) [inline test.rb:20]: Is that a leftover? @joseph-robertson

File: src/model/Model.cpp [hunk 2]
```diff
@@ -742,6 +959,38 @@ namespace model {
...
+      boost::optional<ConvergenceLimits> convergenceLimits = model().convergenceLimits();
+      boost::optional<ShadowCalculation> shadowCalculation = model().shadowCalculation();
+
+      for (const WorkspaceObject& componentObject : component.objects()) {
+        if (componentObject.iddObject().type() == IddObjectType::Building && building) {
+          LOG(Error, "Building");
+        } else if (componentObject.iddObject().type() == IddObjectType::Site_WaterMainsTemperature && siteWaterMainsTemperature) {
+          siteWaterMainsTemperature.get().remove();
+        }
+      }
```
Comment (jmarrec) [inline src/model/Model.cpp:992]: Is it finished or a WIP?

Comment (joseph-robertson) [reply src/model/Model.cpp:992]: WIP

File: src/model/Model.cpp [hunk 3]
```diff
@@ -742,12 +1280,393 @@ namespace model {
...
+              externalInterface->setString(i, componentObject.getString(i).get());
+            }
+          } else {
+            workspaceObjects.push_back(componentObject);
+          }
+        } else {
+          // not a unique model object
+          workspaceObjects.push_back(componentObject);
+        }
+      }
```
Comment (jmarrec) [inline src/model/Model.cpp:1661]: Replace everything with this:

```suggestion
      // Start by deleting existing Unique ModelObject, without touching children/resources by calling WorkspaceObject::remove
      for (const WorkspaceObject& componentObject : component.objects()) {
        auto compIddObject = componentObject.iddObject();
        if (compIddObject.properties().unique) {
          // Remove any existing objects (there should really be only one)
          for (WorkspaceObject& wo : model().getObjectsByType(compIddObject)) {
            LOG(Info, "Removing existing UniqueModelObject in the target model: " << wo.briefDescription());
            wo.remove();
          }
        }
      }
```

**Comment by jmarrec** [comment]
@tijcolem I think CI is largely down. Windows runner for eg has been queuing stuff for the past 1.5 days

File: src/model/Model.cpp [hunk 4]
```diff
@@ -742,12 +1280,26 @@ namespace model {
...
+        auto compIddObject = componentObject.iddObject();
+        if (compIddObject.properties().unique) {
+          // Remove any existing objects (there should really be only one)
+          for (WorkspaceObject& wo : model().getObjectsByType(compIddObject)) {
+            LOG(Info, "Removing existing UniqueModelObject in the target model: " << wo.briefDescription());
+            wo.remove();
+          }
+        }
+      }
+
```
Comment (jmarrec) [inline src/model/Model.cpp:1295]: Actual fix is there.

File: src/model/test/Component_GTest.cpp [hunk 5]
```diff
@@ -463,3 +467,98 @@ TEST_F(ModelFixture, Component_Purge_NonResourceObject) {
...
+    EXPECT_EQ(1u, ocd->numComponentObjects());
+    ASSERT_TRUE(ocd->primaryComponentObject().optionalCast<SiteWaterMainsTemperature>());
+
+    // check model
+    EXPECT_EQ(1u, model.getModelObjects<SiteWaterMainsTemperature>().size());
+    ASSERT_EQ(1u, model.getModelObjects<ComponentData>().size());
+    EXPECT_EQ(ocd->handle(), model.getModelObjects<ComponentData>()[0].handle());
+    EXPECT_EQ(2u, model.numObjects());
+    EXPECT_NE(siteWaterMainsTemperature1.handle(), model.getUniqueModelObject<SiteWaterMainsTemperature>().handle());  // removed
+    EXPECT_NE(siteWaterMainsTemperature2.handle(), model.getUniqueModelObject<SiteWaterMainsTemperature>().handle());  // cloned
```
Comment (jmarrec) [inline src/model/test/Component_GTest.cpp:531]: Please extend the test to to make sure that the resulting object does have the right fields.

File: src/model/Model.hpp [hunk 6]
```diff
@@ -526,24 +686,120 @@ namespace model {
...
+  MODEL_API OutputDiagnostics Model::getUniqueModelObject<OutputDiagnostics>();
+
+  template <>
+  MODEL_API OutputDebuggingData Model::getUniqueModelObject<OutputDebuggingData>();
+
+  template <>
+  MODEL_API OutputJSON Model::getUniqueModelObject<OutputJSON>();
+
+  template <>
+  MODEL_API OutputEnergyManagementSystem Model::getUniqueModelObject<OutputEnergyManagementSystem>();
```
Comment (jmarrec) [inline src/model/Model.hpp:702]: I'm going to be annoying here, sorry. Adding cached getters for all unique model objects is nice, but I think we want to put that in a separate PR. I have a very strong feeling that we'll need partial classes adjustments in C#.

File: src/model/Model_Impl.hpp [hunk 7]
```diff
@@ -161,14 +213,126 @@ namespace model {
...
+     *  object which can be significantly faster than calling getOptionalUniqueModelObject<Site>(). */
+      boost::optional<Site> site() const;
+
+      /** Get the SiteGroundReflectance object if there is one, this implementation uses a cached reference to the SiteGroundReflectance
+     *  object which can be significantly faster than calling getOptionalUniqueModelObject<SiteGroundReflectance>(). */
+      boost::optional<SiteGroundReflectance> siteGroundReflectance() const;
+
+      /** Get the SiteWaterMainsTemperature object if there is one, this implementation uses a cached reference to the SiteWaterMainsTemperature
+     *  object which can be significantly faster than calling getOptionalUniqueModelObject<SiteWaterMainsTemperature>(). */
+      boost::optional<SiteWaterMainsTemperature> siteWaterMainsTemperature() const;
```
Comment (jmarrec) [inline src/model/Model_Impl.hpp:246]: Here's one example getter you added (for caching purposes)

File: src/model/ModelCore.i [hunk 8]
```diff
@@ -27,13 +27,12 @@
...
+  %ignore openstudio::model::Model::outputControlFiles;
   %ignore openstudio::model::Model::outputTableSummaryReports;
+  %ignore openstudio::model::Model::performancePrecisionTradeoffs;
   %ignore openstudio::model::Model::runPeriod;
-  %ignore openstudio::model::Model::weatherFile;
   %ignore openstudio::model::Model::yearDescription;
-  %ignore openstudio::model::Model::performancePrecisionTradeoffs;
-
-  %ignore openstudio::model::Model::outputControlFiles;
+  %ignore openstudio::model::Model::weatherFile;
```
Comment (jmarrec) [inline src/model/ModelCore.i:35]: If I take `SiteWaterMainsTemperature` for eg, it's swigged in ModelSimulation.i too.

so `%ignore openstudio::model::Model::SiteWaterMainsTemperature;` is needed for C# to swig correctly.

**Review by jmarrec — CHANGES_REQUESTED** [review]
Move the cached getter feature to another PR, where we can work on getting C# partial classes working.

I know you originally added them because it was part of your initial solution, but it's now no longer the case after the refactor, so both features can be separated.

File: src/model/Model.hpp (hunk 6 above)
Comment (joseph-robertson) [reply src/model/Model.hpp:702]: Moved to https://github.com/NREL/OpenStudio/pull/4604

**Review by jmarrec — APPROVED** [review]
Thanks for separating it in two PRs!
